### DISCUSSION TITLE
Need help with channel packing

### DISCUSSION THREAD

**Discussion by bjoseph436** [ImageMagick/ImageMagick#2477] [Help] [answered]
I'm a 3D generalist looking to automate channel packing my textures for use in Unreal, this is what i came up with. There are repeating a lot of similar commands, maybe there is a way to make it more efficient, does anyone have any insight on how I can improve this?  Thanks!!

`convert Brick_Wall_Stylized_ao.psd -set colorspace RGB -flatten -colorspace Gray -channel 0 -separate +channel \`
 `\( Brick_Wall_Stylized_roughness.psd -set colorspace RGB -flatten -colorspace Gray -channel 0 -separate +channel \) \`
` \( Brick_Wall_Stylized_albedo.psd -set colorspace RGB -flatten -colorspace Gray -channel 0 -evaluate set 0 -separate +channel \) \`
` \( Brick_Wall_Stylized_albedo.psd -set colorspace RGB -flatten -colorspace Gray -channel 0 -separate +channel \) \`
`-set colorspace RGB -combine -alpha set -colorspace RGB combine_123.png`

*edit formating

**Comment by fmw42** [comment, marked as answer]
Please use the <> button after highlighting your code so that it is properly formatted for line breaks and thus much more readable.  I am puzzled. It looks like you are taking 4 PSD files and extracting the grayscale channel (the last is set to black) and then trying to combine as RGB. First, you cannot combine 4 channels as RGB. It only allows 3. If you want an alpha channel, then you should use -alpha off -compose copy_opacity -composite to add the 4th to the RGB result into the alpha channel.  If you use -colorspace RGB you will get a linear RGB. Is that what you want. If not, then use -colorspace sRGB. If you use the same image more than once, then you should read the input and use either -clone X or use in-memory MPR:X format. That way you can avoid multiple reads of the same file.

**Reply by bjoseph436** [reply]
Thanks for responding...

> Please use the <> button after highlighting your code so that it is properly formatted for line breaks and thus much more readable.

Hopefully fixed in my my original post...

>  I am puzzled. It looks like you are taking 4 PSD files and extracting the grayscale channel (the last is set to black) and then trying to combine as RGB. First, you cannot combine 4 channels as RGB. It only allows 3. If you want an alpha channel, then you should use -alpha off -compose copy_opacity -composite to add the 4th to the RGB result into the alpha channel.

Yes you are correct, I am taking multiple grayscale image and placing it in each channel. Although I do seem to get what I am after with my approach, definitely will try compositing. This opens up a new train of thought!

> If you use -colorspace RGB you will get a linear RGB. Is that what you want. If not, then use -colorspace sRGB. 

Yes this is exactly what I want, the original files are generated as 16-bit linear RGB. I would like to keep the linear RGB while converting to 8-bit png files.  I noticed the images got darker if they were set to -colorspace sRGB.

> If you use the same image more than once, then you should read the input and use either -clone X or use in-memory MPR:X format. That way you can avoid multiple reads of the same file.

What I am trying to do is just fill the G, and Alpha channels with black and white (my offline commands where testing that right before i made the post ) respectively for this particular texture. I found it easiest to just use an image I already had, can the color fill be generated by ImageMagick? I just started using ImageMagick on Tuesday and thank you for the suggestions, i'm curious about this in-memory MPR:X format, will be reading up on it.

**Reply by fmw42** [reply]
I do not think you can fill the alpha channel using -combine. You can make the alpha channel opaque when you combine the RGB channels by simply using -alpha set at the end or using PNG32:outputname as the output. If you want it fully transparent then use -alpha transparent at the end.

**Reply by bjoseph436** [reply]
Awesome, thanks! I will keep these in mind.

**Reply by bjoseph436** [reply]
Thanks for your help! The mpr:x was a big help. I ended up with a file that has RGB channels plus an alpha.

```
convert -size 2048x2048 \
    *file_ao.psd  -colorspace Gray -flatten -write mpr:AO +delete \
    *file_roughness.psd   -colorspace Gray -flatten -write mpr:ROUGHNESS +delete \
    xc:black -write mpr:METALNESS +delete \
    xc:white -write mpr:ALPHA +delete \
    mpr:AO mpr:ROUGHNESS mpr:METALNESS -combine -write mpr:RGBCOMP +delete \
    mpr:RGBCOMP mpr:ALPHA -alpha off -compose copy_opacity -composite -depth 8 png32:filename.png
```

**Comment by snibgo** [comment]
> -colorspace Gray -channel 0 -separate +channel

What is this for? After "-colorspace Gray", there is only one channel. Separating that channel does nothing.

**Reply by bjoseph436** [reply]
Now that you mention, it shouldn't be there, thanks! I have been experimenting a lot thing may be a little dirty...
